### PR TITLE
Fix an issue of too many small steps when msc is enabled

### PR DESCRIPTION
**Pull request by whokion**
Sorry for reopening this MR again, which includes fixes for too many small steps when msc is enabled:
- Add an input argument, geo_limited to UrbanMscScatter and do not to recalculate the true path when the proposed step is not limited by geometry
- Add eloss_calc_limit to PhysicsParamsScalars as a configurable parameter and impose conditional checks for calc_energy_loss
- Adjust expected values in UrbanMsc.test due to changes in UrbanMscScatter

### CODE REVIEW

**Review by sethrj — COMMENTED** [review]
Replaces #451 . One minor issue and one clarification

File: test/celeritas/phys/Physics.test.cc
Comment (sethrj) [inline test/celeritas/phys/Physics.test.cc:125]:
```suggestion
            R"json({"models":[{"label":"mock-model-5","process":0},{"label":"mock-model-6","process":0},{"label":"mock-model-7","process":1},{"label":"mock-model-8","process":2},{"label":"mock-model-9","process":2},{"label":"mock-model-10","process":2},{"label":"mock-model-11","process":3},{"label":"mock-model-12","process":3},{"label":"mock-model-13","process":4},{"label":"mock-model-14","process":4},{"label":"mock-model-15","process":5}],"options":{"eloss_calc_limit":0.001,"enable_fluctuation":true,"energy_fraction":0.8,"fixed_step_limiter":0.0,"linear_loss_limit":0.01,"scaling_fraction":0.2,"scaling_min_range":0.1},"processes":[{"label":"scattering"},{"label":"absorption"},{"label":"purrs"},{"label":"hisses"},{"label":"meows"},{"label":"barks"}],"sizes":{"integral_xs":8,"model_groups":8,"model_ids":11,"process_groups":4,"process_ids":8,"reals":196,"value_grid_ids":75,"value_grids":75,"value_tables":43}})json",
```
Also thinking about it, maybe there's a better description for the parameter than `eloss_calc_limit`? This is essentially a hard cutoff for particles that have energy loss processes. Maybe `eloss_cutoff`?

File: src/celeritas/phys/PhysicsParamsOutput.cc
Comment (whokion) [reply src/celeritas/phys/PhysicsParamsOutput.cc]: Thanks. Modified the option as suggested.

File: test/celeritas/em/UrbanMsc.test.cc
Comment (whokion) [reply test/celeritas/em/UrbanMsc.test.cc]: Added the annotation.

File: test/celeritas/phys/Physics.test.cc
Comment (whokion) [reply test/celeritas/phys/Physics.test.cc]: I agree that eloss_calc_limit is not a good description, but `eloss_cutoff` is not a good one either as it may refer to something that we actually cutoff for the energy loss.  The better description should be something close to "the lowest kinetic energy for the energy loss **calculation**".   Can we leave it as it is for now?

Comment (sethrj) [reply test/celeritas/phys/Physics.test.cc]: Sure, leave it in for now.

**Comment by amandalund** [comment]
@whokion I can get it to crash locally by enabling msc in `TestEm3Base`.. not sure what's going on though (also its looks like it was [crashing before](https://cloud.cees.ornl.gov/jenkins-ci/blue/organizations/jenkins/Celeritas/detail/PR-454/1/tests) too).

With msc enabled + orange it's not crashing but I'm running into some assertion failures related to boundary crossing:
```console
/home/alund/celeritas_project/celeritas/src/orange/OrangeTrackView.hh:363: void celeritas::OrangeTrackView::cross_boundary(): block: [24,0,0], thread: [248,0,0] Assertion `this->surface_id()` failed.
```
It looks like this is because we're calling `move_internal` on geometry-limited tracks (even if the displacement is zero), which clears the surface ID. Is it possible for a track to be displaced by msc if the step was limited by a boundary?

**Comment by whokion** [comment]
@amandalund Thanks for looking into this.  I just tried to make celeritas_global_Stepper crashed with msc turned on in TestEm3Base, but it still runs fine on my machine  - probably the input primary particle in this test may not cover all different cases.  I will play more varying input particles to reproduce the crash locally.

In principle, the msc lateral displacement for the geometry limited steps should not be happened as the safety check in the msc step limit should protect it.  I will double check on that.

**Comment by sethrj** [comment]
Ooh, damn, good debugging work @amandalund .

**Comment by whokion** [comment]
Can you explain why this MR affects the `celeritas_global_Stepper` test (at least on jenkins) as the enable_msc is `false` by default in TestEm3Base?  Also, why does only TestEm3Test.device fail, but not TestEm3Test.host?

**Comment by sethrj** [comment]
Is it possible that it's the "on the boundary with zero energy" error correction causing the problems? You're right that msc should be disabled on device.

One reason it might fail on device but not host is the different number of tracks being used. Since this is clearly a stochastic failure, we might not see it until we run more primaries.

**Comment by amandalund** [comment]
Could also be a bug somewhere else that's just showing up here now. I actually see the crash on master too (with debug build + vecgeom + msc), and with just a single primary...

File: src/celeritas/global/detail/AlongStepActionImpl.hh
Comment (amandalund) [inline src/celeritas/global/detail/AlongStepActionImpl.hh:211]: I think we also need to check that this is a particle with an energy loss process so we don't apply this cut to photons.

Comment (whokion) [reply src/celeritas/global/detail/AlongStepActionImpl.hh:211]: You are absolutely right! Probably we want to move this threshold check back into the calc_energy_loss (without the geo_limited condition?).  Actually, this lowest energy cutoff reduced around 1/3 of number of steps before fixing the true_path, but did not make much difference after the fix (i.e, order dependent) as you did not see much impact of the approximation.  So, another choice may be to move back to the original implementation without any conditional check.

File: src/celeritas/global/detail/AlongStepActionImpl.hh
Comment (sethrj) [inline src/celeritas/global/detail/AlongStepActionImpl.hh:177]: This still isn't quite right, since some cases in the MscScatter (such as being very close to the boundary) still result in a zero displacement even when the earlier step length allows it. The change in #455 to return a new "action result" should hopefully fix the error for all occurrences.

File: src/celeritas/global/detail/AlongStepActionImpl.hh
Comment (amandalund) [reply src/celeritas/global/detail/AlongStepActionImpl.hh]: We do want to keep the geo-limited condition to prevent tracks from stopping on a boundary, right?

File: src/celeritas/em/distribution/UrbanMscStepLimit.hh
Comment (sethrj) [inline src/celeritas/em/distribution/UrbanMscStepLimit.hh:143]: I think this is the third time you've submitted a PR deleting this assertion 😂

I know you wrote on slack that these numbers are very slightly different, but this assertion is there to catch that difference so we can track it down! If it's because we're recalculating the range, we should figure out if caching will help eliminate the difference and improve performance, etc. If the assertion is getting in the way of further analysis, instead of deleting, loosen it a bit (`CELER_ENSURE(range_ >= phys_step_ || soft_equal(range_, phys_step_))`), add a comment, and make sure that nothing else in the class fails when this restriction is broken.

**Comment by sethrj** [comment]
@whokion The new MSC tests are showing a 4x reduction in the number of average steps per primary :) The MSC scatter results that previously were "change direction, no displacement" in the unit tests are now "no change direction, no displacement"—so I think that's why the Stepper + MSC tests are now passing on ORANGE (the "change direction while crossing boundary" case no longer seems to be happening commonly).

File: src/celeritas/em/distribution/UrbanMscStepLimit.hh
Comment (whokion) [reply src/celeritas/em/distribution/UrbanMscStepLimit.hh:143]: Since I did not fully agree the assume that msc is invoked after the eloss range limitation, i.e, the current demo workflow, I still think that it is better to remove this hard restriction here- I mentioned a case for the msc process without the ionization (or bream) process itself, which is also a perfect fine application or test, but you want to ignore anything but the current workflow.  Anyway, we probably do not want to waste our time to argue about potential exceptions or general extensions for the user cases of msc.  Since you refactored this from the original code, I leave to introduce the better solution (probably the caching the range) after merging this MR wihtout any clutch from testem3 test by this assertion.

File: src/celeritas/global/detail/AlongStepActionImpl.hh
Comment (whokion) [reply src/celeritas/global/detail/AlongStepActionImpl.hh]: Okay, a good point.  Yes, it is a side effect by the refactoring, updating geo outside actual place inside the scatter.  Anyway, let's use a corresponding msc action result for position changed that you introduced (in your last MR) in a following MR.  Just to be consistent, we also pass the const reference of the geometry view (instead of the point) as we does not update geometry inside the scatter after the refactoring code.

**Comment by whokion** [comment]
> @whokion The new MSC tests are showing a 4x reduction in the number of average steps per primary :) The MSC scatter results that previously were "change direction, no displacement" in the unit tests are now "no change direction, no displacement"—so I think that's why the Stepper + MSC tests are now passing on ORANGE (the "change direction while crossing boundary" case no longer seems to be happening commonly).

Interesting.! Which means that the condition for skip_sampling is changed for the case happening.  Hopefully this is a true improvement, but not an unwanted by pass.

File: src/celeritas/global/detail/AlongStepActionImpl.hh
Comment (whokion) [reply src/celeritas/global/detail/AlongStepActionImpl.hh]: Not sure whether we still need this extra safe guard - I successfully tested demo-loop with testem3-flat and 100k input file without any problem (and of course, simple-cms  and the hepmc3 input).  So, let's drop it for now until any exceptional case is happened or do you have a new failure because of dropping the condition?

Comment (amandalund) [reply src/celeritas/global/detail/AlongStepActionImpl.hh]: ok, let's leave it as Is :)

Comment (amandalund) [reply src/celeritas/global/detail/AlongStepActionImpl.hh]: (This would only show up as an assertion failure with orange and if the stopped particle was a positron, when the secondary photons try to initialize on a surface.)

For reference, looking at testem3 with 100k primaries, msc enabled, and with this cut, there are ~500 tracks coming to rest on a boundary (with or without energy loss fluctuations, but slightly more with). Without fluctuations only a handful of those are positrons; with fluctuations ~20% are.

Comment (sethrj) [reply src/celeritas/global/detail/AlongStepActionImpl.hh]: Yuck, that's not good... is there an expedient solution? I'm also in the testing phase of the along-step refactor (about to open up a draft pull request for comment now) so we may be able to more cleanly add this cutoff with the new structure?

Comment (amandalund) [reply src/celeritas/global/detail/AlongStepActionImpl.hh]: Not sure what to do with the stopped particles on a boundary from eloss fluctuations, so yeah let's leave it as is here and work on it in your along-step refactor.

**Comment by sethrj** [comment]
@whokion It's likely due to the safety calculation: if VecGeom sees a negative we will return a zero safety which will of course lead to more steps. You could add an `if` block around it or loosen the tolerance until we debug it? Or we could dig in as part of this mr...

**Comment by whokion** [comment]
@sethrj So, do you want me to update results (for is_ci_build() with orange)? - just to double check.

**Comment by mrguilima** [comment]
> @sethrj In my test, there is a difference between vecgoem and orange for the expect number at [Stepper.test.cc:269](https://github.com/whokion/celeritas/blob/fix-msc-rc/test/celeritas/global/Stepper.test.cc#L269) `EXPECT_SOFT_NEAR(55, result.calc_avg_steps_per_primary(), 0.25);` vecgeom Actual 78.375 orange Actual 54.375 Any idea for this difference?

**Comment by mrguilima** [comment]
> vecgeom Actual 78.375
> orange Actual 54.375
> Any idea for this difference?
This is interesting. What are the coordinates at that point?  And which geometry?

**Comment by sethrj** [comment]
@mrguilima This is with TestEm3, the average over all steps over all tracks. I'm adding a unit test that should allow us more detailed exploration of the behavior changes for different test cases, different implementations — but it might be possible to get more data out of the stepper host test without too much trouble.

**Comment by sethrj** [comment]
@whokion yes please if you would. You can add an `if (!CELERITAS_USE_VECGEOM)` around the test of average steps.

File: src/celeritas/global/detail/AlongStepActionImpl.hh
Comment (sethrj) [inline src/celeritas/global/detail/AlongStepActionImpl.hh:155]: We should probably revisit how the "is_displaced" and "geo_limited" and "action result" complement each other: they're not orthogonal choices now.

Comment (whokion) [reply src/celeritas/global/detail/AlongStepActionImpl.hh:155]: Can I leave this to you once you finish refactoring the along step action?

File: src/celeritas/em/distribution/UrbanMscStepLimit.hh
Comment (sethrj) [reply src/celeritas/em/distribution/UrbanMscStepLimit.hh:143]: Thanks!

File: src/celeritas/global/detail/AlongStepActionImpl.hh
Comment (sethrj) [reply src/celeritas/global/detail/AlongStepActionImpl.hh:155]: Yep! Just sort of a note to self.

**Review by sethrj — APPROVED** [review]

**Pull request merged**
